python/feathub/registries/mysql_registry.py

```diff
@@ -0,0 +1,273 @@
+class MySqlRegistryConfig(RegistryConfig):
+    def __init__(self, props: Dict[str, Any]) -> None:
+        super().__init__(props)
+        self.update_config_values(mysql_registry_config_defs)
+
+
+def _get_digest(json_string: str) -> str:
+    return sha256(json_string.encode("utf-8")).hexdigest()
+
+
+class MySqlRegistry(Registry):
+    """
+    A registry that stores entities in a MySQL database.
+    """
+
+    def __init__(
+        self,
+        props: Dict,
+    ):
+        super().__init__("mysql", props)
+        mysql_registry_config = MySqlRegistryConfig(props)
+        self.database = mysql_registry_config.get(DATABASE_CONFIG)
+        self.table = mysql_registry_config.get(TABLE_CONFIG)
+        self.host = mysql_registry_config.get(HOST_CONFIG)
+        self.port = mysql_registry_config.get(PORT_CONFIG)
+        self.username = mysql_registry_config.get(USERNAME_CONFIG)
+        self.password = mysql_registry_config.get(PASSWORD_CONFIG)
+
+        # dict that acts as a local cache for built and registered tables.
+        # each value in the dict is a tuple of table before build and table after build.
+        self.tables: Dict[str, Tuple[TableDescriptor, TableDescriptor]] = {}
+
+        self.conn = mysql.connector.connect(
+            host=self.host,
+            port=self.port,
+            user=self.username,
+            password=self.password,
+            database=self.database,
+        )
+        self.cursor = self.conn.cursor()
+
+        self.cursor.execute(
+            f"""
+                CREATE TABLE IF NOT EXISTS `{self.table}`(
+                   `digest` VARCHAR(64) NOT NULL,
+                   `name` TEXT NOT NULL,
+                   `timestamp` TIMESTAMP NOT NULL,
+                   `is_deleted` BOOLEAN NOT NULL,
+                   `original_json` TEXT NOT NULL,
+                   `built_json` TEXT NOT NULL,
+                   PRIMARY KEY ( `digest`, `timestamp` )
+                );
+            """
+        )
+
+    def build_features(
+        self,
+        feature_descriptors: List[TableDescriptor],
+        force_update: bool = False,
+        props: Optional[Dict] = None,
+    ) -> List[TableDescriptor]:
+        result = []
+        for table in feature_descriptors:
+            if table.name == "":
+                raise FeathubException(
+                    "Cannot build a TableDescriptor with empty name."
+                )
+            self.tables[table.name] = (
+                table,
+                table.build(self, force_update=force_update, props=props),
+            )
+            result.append(self.tables[table.name][1])
+
+        return result
+
+    def register_features(
+        self, feature_descriptors: List[TableDescriptor], force_update: bool = False
+    ) -> bool:
+        self.build_features(feature_descriptors, force_update=force_update)
+
+        for descriptor in feature_descriptors:
+            existing_features = self._get_feature_from_remote_registry_or_none(
```

Comment:
`build_features(` invoked above should have made sure that the given list of `feature_descriptors` can be found in `self.tables`. Would it be simpler to read `self.tables` directly without invoking `_get_feature_from_remote_registry_or_none()` here?